File: problems/problemset/rate-limiter/Python/solution/rate_limiter.py
from collections import deque

class RateLimiter:
    def __init__(self, max_requests: int, window_ms: int):
        self.max_requests = max_requests
        self.window_ms = window_ms
        self.window: deque[int] = deque()

    def allow(self, timestamp: int) -> bool:
        cutoff = timestamp - self.window_ms
        while self.window and self.window[0] < cutoff:
            self.window.popleft()
        if len(self.window) < self.max_requests:
            self.window.append(timestamp)
            return True
        return False

File: problems/problemset/rate-limiter/Rust/solution/rate_limiter.rs
use std::collections::VecDeque;

pub struct RateLimiter {
    max_requests: u32,
    window_ms: u64,
    window: VecDeque<u64>,
}

impl RateLimiter {
    pub fn new(max_requests: u32, window_ms: u64) -> Self {
        Self { max_requests, window_ms, window: VecDeque::new() }
    }

    pub fn allow(&mut self, timestamp: u64) -> bool {
        let cutoff = timestamp.saturating_sub(self.window_ms);
        while self.window.front().map_or(false, |&t| t < cutoff) {
            self.window.pop_front();
        }
        if self.window.len() < self.max_requests as usize {
            self.window.push_back(timestamp);
            true
        } else {
            false
        }
    }
}
